imap/id.go

```diff
@@ -0,0 +1,146 @@
+package imap
+
+import (
+	"context"
+	"encoding/gob"
+	"fmt"
+	"runtime"
+	"strings"
+
+	"github.com/ProtonMail/gluon/internal"
+)
+
+const (
+	IdKeyName             = "name"
+	IdKeyVersion          = "version"
+	IdKeyOS               = "os"
+	IdKeyOSVersion        = "os-version"
+	IdKeyVendor           = "vendor"
+	IdKeySupportUrl       = "support-url"
+	IdKeyAddress          = "address"
+	IdKeyDate             = "date"
+	IdKeyCommand          = "command"
+	IdKeyArguments        = "arguments"
+	IdKeyEnvironment      = "environment"
+	IMAPIDConnMetadataKey = "rfc2971-id"
+)
+
+// ID represents the RFC 2971 IMAP ID Extension. This information can be retrieved by the connector at the context
+// level. To do so please use the provided GetIMAPIDFromContext() function.
+type ID struct {
+	Name        string
+	Version     string
+	OS          string
+	OSVersion   string
+	Vendor      string
+	SupportUrl  string
+	Address     string
+	Date        string
+	Command     string
+	Arguments   string
+	Environment string
+	Other       map[string]string
+}
+
+func NewID() ID {
+	return ID{
+		Name:        "Unknown",
+		Version:     "Unknown",
+		OS:          "Unknown",
+		OSVersion:   "Unknown",
+		Vendor:      "Unknown",
+		SupportUrl:  "",
+		Address:     "",
+		Date:        "",
+		Command:     "",
+		Arguments:   "",
+		Environment: "",
+		Other:       make(map[string]string),
+	}
+}
+
+func (id *ID) String() string {
+	var values []string
+
+	writeIfNotEmpty := func(key string, value string) {
+		if len(value) != 0 {
+			values = append(values, fmt.Sprintf(`"%v" "%v"`, key, value))
+		}
+	}
+
+	writeIfNotEmpty(IdKeyName, id.Name)
+	writeIfNotEmpty(IdKeyVersion, id.Version)
+	writeIfNotEmpty(IdKeyOS, id.OS)
+	writeIfNotEmpty(IdKeyOSVersion, id.OSVersion)
+	writeIfNotEmpty(IdKeyVendor, id.Vendor)
+	writeIfNotEmpty(IdKeySupportUrl, id.SupportUrl)
+	writeIfNotEmpty(IdKeyAddress, id.Address)
+	writeIfNotEmpty(IdKeyDate, id.Date)
+	writeIfNotEmpty(IdKeyCommand, id.Command)
+	writeIfNotEmpty(IdKeyArguments, id.Arguments)
+	writeIfNotEmpty(IdKeyEnvironment, id.Environment)
+
+	for k, v := range id.Other {
+		writeIfNotEmpty(k, v)
+	}
+
+	return fmt.Sprintf("(%v)", strings.Join(values, " "))
+}
+
+func NewIMAPIDFromKeyMap(m *map[string]string) ID {
```

Comment:
map types are already pointer-like. No need to pass them in as a pointer, that's not very idiomatic and really doesn't gain you any performance.

Reply:
Fixed.